### ISSUE TITLE
Add new architecture support: MIPS MSA

### ISSUE THREAD

**Issue by markos** [VectorCamp/vectorscan#156]

**Comment by markos** [comment]
MSA is irrelevant and has no interest whatsoever, the only MSA capable platform in our possession is incredibly buggy and unstable, this is too much work and without any actual benefit. Closing this wishlist bug.

**Comment by FlorianHeigl** [comment]
Hi @markos 

can I ask what the buggy and unstable platform is?


I'm generally more looking into making the MIPS Octeon "HFA" (Hyper Finite Automata) regex things usable, and if that were ever to happen it would probably be best to be compatible wih vectorscan so any interfaces with things like Snort would not require code useless to anyone else. I've seen that Marvell has created opensource support for the newer CN10K series, I'm working with the old stuff, mosty the the 25GbE cards that are now like $30.

But just asking out of curiosity, maybe it's something where i'll hit the bugs myself some day... (Or maybe it's something that is missing in my zoo)

**Comment by markos** [comment]
@FlorianHeigl  I don't have the system here right now, but it is a Loongson 3a3000 or 3a4000. It was before Loongson moved to their own ISA and I managed to make it boot Debian on it. MSA (MIPS SIMD Architecture) extention was reported in the cpu features. However, it was constantly crashing and it was impossible to stay online for more than a few hours. It could be anything really and given that there was zero interest and funding for the platform, I gave up trying to make it work. Unfortunately, even if someone provides the patches for this, I will not merge them as there is no platform to test. 
Unless I do manage to get one of those boards that you say for $30, maybe I could revisit that idea. Could you point me to such a board?

**Comment by FlorianHeigl** [comment]
Hi,

Firstofall, wow, at least it's an exotic crashing thing then ;=)

What I'm playing with is not Loongson and I think it doesn't have vector instructions. I came here looking into it.
This is the flags I see on one of my test systems:

```
processor		: 15
cpu model		: Cavium Octeon III V0.3  FPU V0.0
BogoMIPS		: 3600.00
wait instruction	: yes
microsecond timers	: yes
tlb_entries		: 256
extra interrupt vector	: yes
hardware watchpoint	: yes, count: 2, address/irw mask: [0x0ffc, 0x0ffb]
isa			: mips1 mips2 mips3 mips4 mips5 mips64r2
ASEs implemented	: vz
shadow register sets	: 1
kscratch registers	: 4
core			: 15
VCED exceptions		: not available
VCEI exceptions		: not available
```


I've picked one example card for you on ebay US since I don't know where you are at...
[see this offer for a "CN2360"](https://www.ebay.com/itm/206198247342) those are what I try to make generally accessible. (2360 -> 16 core, 8GB. Slow cores, but has offload engines)

I can absolutely help you get it to boot Linux, and show you around the public SDK, but I don't have any knowledge about the offload stuff or how to make those examples really work. I'm not a dev and they (Cavium) had hidden all their info behind NDAs.
Note you need to put a FAN on them, they have a thermal watchdog but that thing only crashes the card again and again instead of clocking it down or something like that. Just keep it cool and avoid the hassle.

What I can say is it'll **not** be using SIMD but some **proprietary** instructions to talk to whatever machinery was on the card. From my understanding, each core had some parallel friends for some offload tasks and other things were shared. There's a copy of a [datasheet for the Octeon CN73xx CPU](https://manuals.plus/m/8d8c34da68e0cdceec739e9a7e99f09916dc82291b274f4b58224b5f800e1cfe)
If you look at the block diagram, where it says "2 HFA Clusters". HFA stands for "Hyper Finite Automata" - that is the thing that would do the (offloaded) work of vectorscan, dealing with Regex and whatevers.
I know there's some examples in the SDK for this stuff and so I consider it _theoretically_ solveable. But that's where I'm at, not further. I am porting Alpine to them, that's my actual goal, along with 50 sidequests. All of it is making progress but all of it is also constantly getting put aside when it gets hard.

I only work on this because I can't do my normal stuff and out of spite because I'm angry people let the hardware become e-waste simply because their lawyers were so deeply infatuated with all their NDAs that they let the whole company and architecture die. But it definitely is an exercise in necromancy what I'm doing here. OTOH there's a few people who dabble with this so that's good.

The more sane target with vectorscan is on the -current- CN10K DPUs
That stuff is much better managed, and more open, i.e. the [DAO code repos here](https://github.com/MarvellEmbeddedProcessors/dao) and the docs here that also [include vectorscan out of the box](https://marvellembeddedprocessors.github.io/dao/guides/dao-devel/applications/snort.html)

I'm working in part on the 'mild' assumption that they reused some hardware logic in their "classification engines" from the original HFA engines. But even if not, then, like I wrote, it would still be best to couple the HFA stuff into vectorscan since it would be definitely possible and be the most pluggable way.

**Comment by markos** [comment]
@FlorianHeigl yeah it doesn't have MSA, so porting just the SIMD part is not possible, but that one can be easily solved by libSIMDe backend. Now, the HFA backend is definitely doable, but it's going to cost time which I don't have. And to be honest it's probably not worth the effort. It would literally be faster and easier to take a FPGA card with dual QSFP28 and write our own custom HFA offloading engine, which is something I tried to get funding for but there was no interest.

Regardingcurrent CN10k DPUs, of course Vectorscan works on the recent Arm-based CPUs, using SIMD and SVE2. However, HFA support for modern CPUs  would be nice to have perhaps as an enterprise/licensed version and provide an even better performance. Same problem as before, it's hard to provide support for a closed architecture, unless someone goes into the trouble of reverse engineering it. Unfortunately, I don't have the resources to do that. 

It's possible Marvell have done their own HFA backend for Vectorscan -possible, but I'm not saying they have done that, I have no idea- and not contributed it back then that's unfortunate.

This is the problem with BSD licensing, people can add their own stuff and never contribute back to the original project. If I could I would relicense the project under LGPL right from the start. It's still possible but harder.

Arm have stopped funding this project so heavy development lies entirely up to me. Occasional contributors send patches, but without funding, I'm not accepting anything complicated, only bug fixes and trivial optimizations, I will never accept big architecture support patches that require extra maintainance effort. I have a huge roadmap of things to fix (like C++26 modernization, removal of Boost, cleaning up SIMD code in old spaghetti C code, etc) and very little help to do that.

As much as I would like to help, the best I can offer is instructions on how to get it run on your platform with the SIMDe backend.

**Comment by FlorianHeigl** [comment]
Hey @markos thanks for the reply. I'm aware of the constraints.
If, supposedly, I'm suddenly smart enough to reverse engineer my card's HFA functionality, I would strive to make patches that do not introduce any (architectural) changes to vectorscan at all. I think that's the plausible approach since it'd only need to skip over certain functions (i.e. building the regex db) or run a different 'thing' for others (idk). but, when considering that I think even a LD_PREOAD approach would work for that; then you'd not need to consider any changes at all because there would be none.
But anyway, let's worry about all that only in case it ever becomes a reality :-)

I'm keeping one sidequest for myself - if i get a hold of one of the CN10k cards, see how/if/what Marvell has changed and push things toward them contributing properly whatever they have made. or at least make them aware they should be providing some partial manpower.
IMO you have already helped by giving so much insight. I hope ARM funds it again at some point, or some of the many security vendors that might or must be profiting from it as well :)

And I'll keep looking for some instruction set docs for these CPUs. If they turn out to really have SIMD instructions then i'll be back for that :-)

Happy hacking!

**Comment by markos** [comment]
@FlorianHeigl the important thing with reverse engineer, is to document what you find. With enough documentation, we could add vectorscan support and possibly other things as well. But it's a really hard task and it will take time. If/when you manage to get even a small subset of these instructions rev.engineered and documented I would be happy to help investigating how we could use them inside Vectorscan.
